### ISSUE TITLE
Scrolling jumping back

### ISSUE THREAD

**Issue by NathanielIR** [mlemgroup/mlem#296]
When scrolling through a long comments section (or sometimes a medium one), It'll get to a point where the app starts jumping back a little as I scroll. So I'll scroll past a comment and then it'll jump back to that comment over and over again and won't let me past unless I grab the side bar and swipe down super quickly

https://github.com/mlemgroup/mlem/assets/88664846/3986ff65-fe71-4ce1-9f52-700f26a2d0e8

**Comment by boscojwho** [comment]
Tested out my hunch that this bug may be caused by nested comments: I created a post on our test instance with various top-level comments (no replies) [See here](https://test-mlem.jo.wtf/post/25).

This bug is not present when a post has many top-level only comments (i.e. no replies).

Will continue to investigate further =)

**Comment by boscojwho** [comment]
Bug is not reproducible when there are only "top-level" comments, but notice here how the scroll bar position **jumps** as we scroll up, even though the content doesn't jump. Interesting.

https://github.com/mlemgroup/mlem/assets/2549615/8bc5c9ca-d9b8-444e-9842-0c3b2d1d2a5e

**Comment by boscojwho** [comment]
Image posts don't appear to be causing this issue: [See here](https://test-mlem.jo.wtf/post/26)

**Comment by boscojwho** [comment]
Can reproduce bug when viewing [this post here](https://lemmy.world/post/1907661).

**Comment by boscojwho** [comment]
My current "hunch" is that the nested `LazyVStack`s in `ExpandedPost` and `CommentItem` child comments aren't making it easy for SwiftUI to do whatever internal calculations it needs to set the overall list view's content size....I say this mostly because of the jumping scroll bar when scrolling up and down.

Will investigate more later today =)

**Comment by NathanielIR** [comment]
Awesome! Thanks for looking into it mate!

**Comment by boscojwho** [comment]
So yes, this bug has to do with the way we are using `LazyVStack`s.
- If we replace both `ExpandedPost` and `CommentItem` to use `VStack`, this bug goes away. Note that bug still persists if we only replace either one of them.
- I realized that `CommentItem` is actually recursively nesting more child instances of itself (and more `LazyVStack`s).

**Comment by boscojwho** [comment]
TL;DR: On scroll up, SwiftUI is triggering redraws on `CommentItem` due to view value and **view identity** change.

Solutions:
1. Fix whatever is causing `CommentItem`'s identity to change.
-- Technically, on scroll up, only the view's value should change, so this is not expected behaviour.  I made a test project with the same basic nested LazyVStack structure, and only the view's value is changed, as expected.

2. Use a single `List` to present all parent/child comments.
-- I already got this working by flattening the hierarchical comments into a single array, then rendering them in a single `List`, and everything works as expected (with the added bonus of much smoother scrolling).
-- This would take a bit more time to implement. We'd have to figure out how to collapse child comments (We won't get this behaviour for free anymore if child comments aren't nested views). Also need to write proper logic for flattening comments.

I'll probably take a little bit more time to figure out Solution 1, but having nested `LazyVStack`s seems a bit janky to me, and not really something it was designed for perhaps?

**Comment by EricBAndrews** [comment]
A bit late to this party, but check out #396 re: identity--we're migrating away from raw API types and to proper middleware models, which should give us an opportunity to handle identity intelligently

**Comment by boscojwho** [comment]
Will have PR ready for this once I get child comments to remember collapsed state.

**Comment by boscojwho** [comment]
See PR #407

**Comment by boscojwho** [comment]
Would have preferred to use `List` over `LazyVStack` for better scrolling performance, but `List` doesn't work inside `ScrollView`, and I didn't want to mess with the view tree anymore for now (i.e. we could in the future put the actual post in a Section header, and render comments as one section in a List).

**Comment by boscojwho** [comment]
@NathanielIR  Squashed this issue, no more portaling back and forth between comments like a wizard =P Fix will be out on a test build *soon*ish.

**Comment by NathanielIR** [comment]
Awesome! Thanks for looking into it mate!

**Comment by EricBAndrews** [comment]
Closed in #407
